File: apps/ui/src/chapterName.ts
// One rule for a chapter's name everywhere it is shown (chapter-title-display-consistency.prd.md): title and subtitle
// join with " — " (space, em dash, space) in source casing, never CSS capitals, never italic or monospace. This is the
// only place that formats one into plain text; primitives/TitleSubtitle.tsx is the only place that renders it as markup.
// Read chapterNameFormatting.test.ts before adding another site that reads `.subtitle` directly.

export interface ChapterNameInput {
  title: string;
  subtitle?: string;
}

// A context prefix ("Read aloud", "Recording check", "Stage suggestion"), built with context(). Kept distinct from
// 'full'/'short' so a caller cannot pass an arbitrary string where a form is expected.
export interface ChapterNameContext {
  readonly prefix: string;
}

export type ChapterNameForm = 'full' | 'short' | ChapterNameContext;

export function context(prefix: string): ChapterNameContext {
  return { prefix };
}

const WHITESPACE = /\s+/g;
// A separator the title already ends with, so appending " — subtitle" cannot double up ("CHAPTER ONE: — Subtitle").
const TRAILING_SEPARATOR = /[:—–-]\s*$/;

function squash(text: string): string {
  return text.replace(WHITESPACE, ' ').trim();
}

// A legacy title that still holds its subtitle after a literal "\n" (three Go/Python helpers used to join these with
// ": " - see chapterName.test.ts), read the same way the importers' headingParts splits a heading: the first line is
// the title, every later line joins into the subtitle.
function splitLegacyNewline(title: string): { title: string; subtitle?: string } {
  if (!title.includes('\n')) return { title: squash(title) };
  const [first, ...rest] = title.split('\n');
  const subtitle = squash(rest.join(' '));
  return { title: squash(first ?? ''), subtitle: subtitle || undefined };
}

function fullName(title: string, subtitle: string | undefined): string {
  if (!subtitle) return title;
  const stripped = title.replace(TRAILING_SEPARATOR, '').trim();
  return stripped ? `${stripped} — ${subtitle}` : subtitle;
}

// The chapter's name as plain text. 'full' (the default) is "Title — Subtitle", or the title alone when there is no
// subtitle. 'short' is always the title alone - for a control inside a row that already shows the full name. A context
// prefix (context('Read aloud')) gives "Prefix: Title — Subtitle".
export function chapterName(chapter: ChapterNameInput, form: ChapterNameForm = 'full'): string {
  const legacy = splitLegacyNewline(chapter.title);
  const title = legacy.title;
  const subtitle = chapter.subtitle !== undefined ? squash(chapter.subtitle) || undefined : legacy.subtitle;

  if (typeof form === 'object') return `${form.prefix}: ${fullName(title, subtitle)}`;
  if (form === 'short') return title;
  return fullName(title, subtitle);
}

File: apps/desktop/internal/chaptername/chaptername.go
// Package chaptername formats a chapter's name as plain text by the app's one rule (chapter-title-display-consistency
// PRD, ADR 0191): the title and the subtitle join with " — " in the book's own casing. It mirrors
// apps/ui/src/chapterName.ts and narration_common.chapter_names; tests/fixtures/chapter-names.json, written by this
// package's test, pins all three to the same output. Plain-text outputs that a consumer may not take an em dash in
// (file names, some tags) use the Plain form, " - " (owner decision D34).
package chaptername

import (
	"regexp"
	"strings"
)

// Form is how a name is written. Full is "Title — Subtitle" (the title alone without a subtitle); Short is the title
// alone; Plain is Full with " - " in place of the em dash. A context prefix is Context("Read aloud").
type Form struct {
	kind   string
	prefix string
}

var (
	Full  = Form{kind: "full"}
	Short = Form{kind: "short"}
	Plain = Form{kind: "plain"}
)

// Context is "Prefix: Title — Subtitle".
func Context(prefix string) Form { return Form{kind: "context", prefix: prefix} }

// trailingSeparator is a separator the title already ends with, so joining a subtitle cannot double up
// ("CHAPTER ONE: — Subtitle").
var trailingSeparator = regexp.MustCompile(`[:—–-]\s*$`)

func squash(text string) string { return strings.Join(strings.Fields(text), " ") }

// splitLegacyNewline reads a title that still holds its subtitle after a line break, as the importers split a heading:
// the first line is the title and every later line joins into the subtitle.
func splitLegacyNewline(title string) (string, string) {
	first, rest, found := strings.Cut(title, "\n")
	if !found {
		return squash(title), ""
	}
	return squash(first), squash(strings.ReplaceAll(rest, "\n", " "))
}

func joined(title, subtitle, separator string) string {
	if subtitle == "" {
		return title
	}
	stripped := strings.TrimSpace(trailingSeparator.ReplaceAllString(title, ""))
	if stripped == "" {
		return subtitle
	}
	return stripped + separator + subtitle
}

// Name writes a chapter's name. subtitle is nil when the chapter has no subtitle field, and then a title holding a
// line break gives its later lines as the subtitle; a subtitle that is set (even empty) is the subtitle.
func Name(title string, subtitle *string, form Form) string {
	name, legacy := splitLegacyNewline(title)
	sub := legacy
	if subtitle != nil {
		sub = squash(*subtitle)
	}
	switch form.kind {
	case "short":
		return name
	case "plain":
		return joined(name, sub, " - ")
	case "context":
		return form.prefix + ": " + joined(name, sub, " — ")
	}
	return joined(name, sub, " — ")
}
